keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -0,0 +1,355 @@
+```
+
+## Sample Case 1: 20,000 endpoints, 5,000 nodes
+ 
+### Service Creation/Deletion
+
+
+|                          | Endpoints             | 100 Endpoints per EndpointSlice | 1 Endpoint per EndpointSlice |
+|--------------------------|-----------------------|---------------------------------|------------------------------|
+| # of writes              | O(1)                  | O(P/B)                          | O(P)                         |
+|                          | 1                     | 200                             | 20000                        |
+| Size of API object       | O(P)                  | O(B)                            | O(1)                         |
+|                          | 20k * const = ~2.0 MB | 100 * const = ~10 KB            | < ~1KB                       |
+| # of watchers per object | O(N)                  | O(N)                            | O(N)                         |
+|                          | 5000                  | 5000                            | 5000                         |
+| # of total watch event   | O(N)                  | O(NP/B)                         | O(NP)                        |
+|                          | 5000                  | 5000 * 200 = 1,000,000          | 5000 * 20000 = 100,000,000   |
+| Total Bytes Transmitted  | O(PN)                 | O(PN)                           | O(PN)                        |
+|                          | 2.0MB * 5000 = 10GB   | 10KB * 5000 * 200 = 10GB        | ~10GB                        |
+
+### Single Endpoint Update
+
+|                          | Endpoints             | 100 Endpoints per EndpointSlice | 1 Endpoint per EndpointSlice |
+|--------------------------|-----------------------|---------------------------------|------------------------------|
+| # of writes              | O(1)                  | O(1)                            | O(1)                         |
+|                          | 1                     | 1                               | 1                            |
+| Size of API object       | O(P)                  | O(B)                            | O(1)                         |
+|                          | 20k * const = ~2.0 MB | 100 * const = ~10 KB            | < ~1KB                       |
+| # of watchers per object | O(N)                  | O(N)                            | O(N)                         |
+|                          | 5000                  | 5000                            | 5000                         |
+| # of total watch event   | O(N)                  | O(N)                            | O(N)                         |
+|                          | 5000                  | 5000                            | 5000                         |
+| Total Bytes Transmitted  | O(PN)                 | O(BN)                           | O(N)                         |
+|                          | ~2.0MB * 5000 = 10GB  | ~10k * 5000 = 50MB              | ~1KB * 5000 = ~5MB           |
```

Comment:
Not sure a single endpoint update is a good thing to think about, since in practice we should expect multiple endpoints to change within the time period in a huge service.

Reply:
I don't agree with that. We want to keep the "batching period" fairly small (less than 1s). Given that, (obviously assuming there isn't a rollout or sth happening now), it's very rare to have changes of more than 1 pod within that period. So I actually think that it is a metric that we should be looking at.

Reply:
Well, then it's not comparing apples to apples-- the current mechanism is like 10s worth of batching IIRC. So when the rate of pod changes is above .1 per second, the rate of updates is a constant .1 / second, and never goes higher.

There's some implicit assumption here about how frequently state-changing things happen to pods. I am not sure what a good assumption is. If we assume the "mean time to state change" is like 1 day (~86k seconds), then we'd expect a service with 20k pods to have about 1 change every 5 seconds, which implies two (small) updates in the new system vs one (big) update in the old system.

But the calculation is pretty sensitive to that time-- if my nodes are worse or my pods are buggier, and the MTtST is only 1 hour, then it's ~5.5 changes per second, and the old mechanism gives 1 big update every 10s (10GB total) and the new system gives ~55 (small) updates (50MB*55=2.7GB) in the same amount of time.

So, the new method is still much better on bandwidth, **but it is by a factor of 4, not a factor of 200**.

With watch events in that latter scenario, the new method causes 5000*55 = ~250k events, compared to 5k with the old method. So it is clearly much worse by this metric. I *think* at these numbers bandwidth is more important than events, but I'm not 100% sure.

Reply:
I think I understand what @lavalamp is saying. 

Here is a worst case scenario: 10k endpoints,  1 big Endpoints object vs. 100 EndpointSlice Objects. 
Within "batching period", let us say 100 endpoints got removed and they happen to distribute across all 100 EndpointSlice objects. The old API and the new API will perform the same in this case as all endpoints have to be redistributed. 



For removal, we can not do much as the endpoints may be scattered.
For addition, the controller can try to pack the new endpoints together and add to one slice.
For noisy pods, the controller can try to concentrate them into a few EndpointSlice. (This seems complex)

Reply:
@lavalamp - I think there is one false assumption in your thing which is: "the current mechanism is like 10s worth of batching".
This is what isn't true today. In EndpointsController (which we're talking about here) we batch "everything that happened until now" - so if pods are changing very fast we will batch some stuff, but it's more like 0.1s rather than 10s.

To clarify a bit what is happening in EndpointsController is we have a queue from which workers are taking services to process, and when processing a given service we list pods from cache and compute the Endpoints object. There is no explicit batching - batching is only a result of lag in controller.
Then 10s batching I guess you might have been talking about batching on kubeproxy side to update iptables.

And I think that explains our disagreement about this point - because we have different assumptions.
Because what I'm saying is that it rare (if we exclude creation or deletion the whole service) to have things batched by endpoints-controller.

Reply:
I must be misremembering something--I thought endpoints controller deliberately batched. IMO it *should* do that, it would be a massive bandwidth savings for high-churn services.

If it doesn't deliberately batch, it'll still be limited by how long it takes to write the endpoints object, but I agree that should be < 1 second even in pretty bad conditions.

Reply:
> IMO it should do that, it would be a massive bandwidth savings for high-churn services.

There is a trade-off with how fast changes are reflected in the iptables (or whatever else is used) [e.g. it is on the path for autoscaling where speed is something that customers are complaining]

So in general, I agree we can be much smarter, but it's also not that obvious.